File: context_tree_cli/verify.py
"""Verify command: run deterministic checks against the tree."""

from __future__ import annotations

import re
import subprocess
import sys

from context_tree_cli.repo import Repo

UNCHECKED_RE = re.compile(r"^- \[ \] (.+)$", re.MULTILINE)


def _check(label: str, passed: bool) -> bool:
    icon = "\u2713" if passed else "\u2717"
    status = "PASS" if passed else "FAIL"
    print(f"  {icon} [{status}] {label}")
    return passed


def _check_progress(repo: Repo) -> list[str]:
    """Read progress.md and return any unchecked items."""
    text = repo.read_file(".context-tree/progress.md")
    if text is None:
        return []
    return UNCHECKED_RE.findall(text)


def run_verify() -> int:
    repo = Repo()
    all_passed = True

    print("Context Tree Verification\n")

    # --- Progress file check ---
    unchecked = _check_progress(repo)
    if unchecked:
        print("  Unchecked items in .context-tree/progress.md:\n")
        for item in unchecked:
            print(f"    - [ ] {item}")
        print()
        print(
            "  Verify each step above and check it off in progress.md"
            " before running verify again.\n"
        )
        all_passed = False

    # --- Deterministic checks ---
    print("  Checks:\n")

    # 1. Framework exists
    all_passed &= _check(
        ".context-tree/VERSION exists",
        repo.has_framework(),
    )

    # 2. Root NODE.md has valid frontmatter
    fm = repo.frontmatter("NODE.md")
    has_valid_node = (
        fm is not None
        and "title" in fm
        and "owners" in fm
    )
    all_passed &= _check(
        "Root NODE.md has valid frontmatter (title, owners)",
        has_valid_node,
    )

    # 3. AGENT.md exists with framework markers
    all_passed &= _check(
        "AGENT.md exists with framework markers",
        repo.has_agent_md_markers(),
    )

    # 4. validate_nodes.py passes
    validate_script = repo.root / ".context-tree" / "validate_nodes.py"
    if validate_script.is_file():
        result = subprocess.run(
            [sys.executable, str(validate_script)],
            capture_output=True,
            text=True,
        )
        all_passed &= _check(
            "validate_nodes.py passes",
            result.returncode == 0,
        )
        if result.returncode != 0 and result.stdout:
            for line in result.stdout.strip().splitlines():
                print(f"    {line}")
    else:
        all_passed &= _check("validate_nodes.py exists", False)

    # 5. At least one member node exists
    all_passed &= _check(
        "At least one member node exists under members/",
        repo.member_count() > 0,
    )

    print()
    if all_passed:
        print("All checks passed.")
    else:
        print("Some checks failed. See above for details.")
    return 0 if all_passed else 1

File: skills/first-tree-cli-framework/references/repo-snapshot/src/verify.ts
import { Repo } from "#src/repo.js";
import { runValidateMembers } from "#src/validators/members.js";
import { runValidateNodes } from "#src/validators/nodes.js";

const UNCHECKED_RE = /^- \[ \] (.+)$/gm;

export function check(label: string, passed: boolean): boolean {
  const icon = passed ? "\u2713" : "\u2717";
  const status = passed ? "PASS" : "FAIL";
  console.log(`  ${icon} [${status}] ${label}`);
  return passed;
}

export function checkProgress(repo: Repo): string[] {
  const text = repo.readFile(".context-tree/progress.md");
  if (text === null) return [];
  const matches: string[] = [];
  let m: RegExpExecArray | null;
  UNCHECKED_RE.lastIndex = 0;
  while ((m = UNCHECKED_RE.exec(text)) !== null) {
    matches.push(m[1]);
  }
  return matches;
}

export interface ValidateNodesResult {
  exitCode: number;
}

export type NodeValidator = (root: string) => ValidateNodesResult;

function defaultNodeValidator(root: string): ValidateNodesResult {
  const { exitCode } = runValidateNodes(root);
  return { exitCode };
}

export function runVerify(repo?: Repo, nodeValidator?: NodeValidator): number {
  const r = repo ?? new Repo();
  const validate = nodeValidator ?? defaultNodeValidator;
  let allPassed = true;

  console.log("Context Tree Verification\n");

  // Progress file check
  const unchecked = checkProgress(r);
  if (unchecked.length > 0) {
    console.log("  Unchecked items in .context-tree/progress.md:\n");
    for (const item of unchecked) {
      console.log(`    - [ ] ${item}`);
    }
    console.log();
    console.log(
      "  Verify each step above and check it off in progress.md before running verify again.\n",
    );
    allPassed = false;
  }

  // Deterministic checks
  console.log("  Checks:\n");

  // 1. Framework exists
  allPassed = check(".context-tree/VERSION exists", r.hasFramework()) && allPassed;

  // 2. Root NODE.md has valid frontmatter
  const fm = r.frontmatter("NODE.md");
  const hasValidNode =
    fm !== null && fm.title !== undefined && fm.owners !== undefined;
  allPassed = check(
    "Root NODE.md has valid frontmatter (title, owners)",
    hasValidNode,
  ) && allPassed;

  // 3. AGENT.md exists with framework markers
  allPassed = check(
    "AGENT.md exists with framework markers",
    r.hasAgentMdMarkers(),
  ) && allPassed;

  // 4. Node validation
  const { exitCode } = validate(r.root);
  allPassed = check("Node validation passes", exitCode === 0) && allPassed;

  // 5. Member validation
  const members = runValidateMembers(r.root);
  allPassed = check("Member validation passes", members.exitCode === 0) && allPassed;

  console.log();
  if (allPassed) {
    console.log("All checks passed.");
  } else {
    console.log("Some checks failed. See above for details.");
  }
  return allPassed ? 0 : 1;
}
